torch/utils/tensorboard/_pytorch_graph.py

```diff
@@ -0,0 +1,267 @@
+        self.valid_methods = valid_methods[:]
+        self.inputs = []
+
+        for m in self.valid_methods:
+            if m == 'inputs' or m == 'outputs':
+                list_of_node = list(getattr(NodeCpp, m)())
+                io_unique_names = []
+                io_tensor_sizes = []
+                for n in list_of_node:
+                    io_unique_names.append(n.uniqueName())
+                    if n.type().kind() == 'CompleteTensorType':
+                        io_tensor_sizes.append(n.type().sizes())
+                    else:
+                        io_tensor_sizes.append(None)
+
+                setattr(self, m, io_unique_names)
+                setattr(self, m + 'tensor_size', io_tensor_sizes)
+
+            else:
+                setattr(self, m, getattr(NodeCpp, m)())
+
+
+class NodePyIO(NodePy):
+    def __init__(self, NodeCpp, input_or_output=None):
+        super(NodePyIO, self).__init__(NodeCpp, methods_IO)
+        try:
+            tensor_size = NodeCpp.type().sizes()
+        except RuntimeError:
+            tensor_size = [1, ]  # fail when constant model is used.
+        self.tensor_size = tensor_size
+        self.kind = 'Parameter'
+        if input_or_output:
+            self.input_or_output = input_or_output
+            self.kind = 'IO Node'
+
+
+class NodePyOP(NodePy):
+    def __init__(self, NodeCpp):
+        super(NodePyOP, self).__init__(NodeCpp, methods_OP)
+        self.attributes = str({k: NodeCpp[k] for k in NodeCpp.attributeNames()}).replace("'", ' ')
+        self.kind = NodeCpp.kind()
+
+
+class GraphPy(object):
+    def __init__(self):
+        self.nodes_op = []
+        self.nodes_io = OrderedDict()
+        self.uniqueNameToScopedName = {}
+        self.shallowestScopeName = 'default'
+        self.scope_name_appeared = []
+
+    def append(self, x):
+        if type(x) == NodePyIO:
+            self.nodes_io[x.uniqueName] = x
+        if type(x) == NodePyOP:
+            self.nodes_op.append(x)
+            for node_output, outputSize in zip(x.outputs, x.outputstensor_size):
+                self.scope_name_appeared.append(x.scopeName)
+                self.nodes_io[node_output] = NodeBase(node_output,
+                                                       x.inputs,
+                                                       x.scopeName,
+                                                       outputSize,
+                                                       op_type=x.kind,
+                                                       attributes=x.attributes)
+
+    def printall(self):
+        print('all nodes')
+        for node in self.nodes_op:
+            print(node)
+        for key in self.nodes_io:
+            print(self.nodes_io[key])
+
+    def find_common_root(self):
+        for fullscope in self.scope_name_appeared:
+            if fullscope:
+                self.shallowestScopeName = fullscope.split('/')[0]
+
+    def populate_namespace_from_OP_to_IO(self):
```

Comment:
I spent a few minutes trying to understand what this function does, and gave up. A quick English description of what this is doing would be very helpful.

Reply:
@lanpa, could I get your help on this description? Thanks.

Reply:
@lanpa can you comment here? Thanks.